### ISSUE TITLE
Rework HTML generation

### ISSUE THREAD

**Issue by acoburn** [trellis-ldp/trellis#73]
The HTML version of resource serializations could use some significant polish, both code-wise and from a UI design perspective. It would also be helpful to clean up the `JenaIOService` constructor as part of this.

**Comment by acoburn** [comment]
Ideally, we would keep the HtmlSerialization code in the IO Service. That keeps the semantics of I/O consistent. In this sense, the relevant interface is:

```java
void IOService::write(Stream<? extends Triple> triples, OutputStream output, RDFSyntax syntax, IRI... profiles);
```

And the HTML output will be generated when the `RDFSyntax` is [RDFA](https://commons.apache.org/proper/commons-rdf/apidocs/index.html?org/apache/commons/rdf/api/package-summary.html)

All of the existing HTML-specific methods are internal, so it is possible to completely rewrite the HTML generation without making any changes to the `http` layer. The current impl uses a mustache template -- in principle, a user can customize this, which is nice, but I'm not wedded to that approach.

Basically, you'd want to start [here](https://github.com/trellis-ldp/trellis/blob/master/trellis-io-jena/src/main/java/org/trellisldp/io/JenaIOService.java#L150) and then rework the code in `./impl`.

**Comment by christopher-johnson** [comment]
A distinct JS "front end" could also be completely decoupled from trellis-app and use the api to get the desired resource serialization.  I tend to favor this approach since it is completely non-invasive.

**Comment by acoburn** [comment]
Yes, a completely JS-based front-end would be really excellent. I also would favor that approach.

**Comment by ajs6f** [comment]
+1 to non-obtrusive. Quality HTML generation is genuinely hard, so it makes sense to parcel it off into a separate concern with some independence.

**Comment by christopher-johnson** [comment]
I see a possibility for this that could facilitate multiple front-end implementations.  The first task would be to document and define the TrellisLDP HTTP API using the [OpenAPI specification](https://github.com/OAI/OpenAPI-Specification).  This API is more complicated than it seems, but it seems mature and widely accepted.

Then a front-end could read this document to create a browser interface.  I experimented with one [such interface](https://koumoul.com/openapi-viewer/?url=https://koumoul.com/s/geocoder/api/v1/api-docs.json&proxy=false).  It would require some modification to fully support TrellisLDP.

**Comment by ajs6f** [comment]
No genuinely RESTful API needs a separate description in a separate IDL (HATEOAS instead), and LDP is genuinely RESTful. I'm not trying to argue against anyone writing an OpenAPI schema for LDP, I'm just (preemptively) arguing against it being included as part of the Trellis project.

I'd rather see stateless front-ends that just pass links. If folks want to write something else, okay, but the burden of maintaining a particular description should stay with them.

**Comment by christopher-johnson** [comment]
A HATEOAS driven front end interface would be brilliant, but it would probably have to built from scratch.  I like the idea better than using OpenAPI actually.  I am in kind of a rush to provide a simple admin interface, though, and I was looking for a short-cut by reusing existing code.  I could probably do this in a day or two using OpenAPI.  A HATEOAS front end would take considerably more work.

**Comment by ajs6f** [comment]
Sorry, @christopher-johnson, I may have miscommunicated. I'm not in any way against you making an OpenAPI-based front end. I'm just saying that we should keep the OpenAPI dependency _in the front end_. IOW, let's don't commit Trellis going forward to maintaining an OpenAPI schema. That was my only concern. But as for making an OpenAPI-based front-end (including keeping the schema in it), go for it, I'm not trying to stop you at all.

**Comment by acoburn** [comment]
I just want to mention that a "JavaScript-based admin interface" has been on the Trellis [development roadmap](https://github.com/trellis-ldp/trellis/wiki/Development-Roadmap#summer-2018) for a while, so I am really looking forward to seeing where this goes. And are you planning to use a modern JS framework like React or Vue for this? I would be delighted to contribute to this effort.

 My only general opinion at this point is: aim to use ES6 and don't use JQuery

**Comment by acoburn** [comment]
@christopher-johnson this looks great. Right now, the trellis code doesn't really support configurable templates (but it should!). I'll make an adjustment to the code, and once that's done, it should be possible to add this template as a classpath resource. That way, simply by adjusting the configuration, a user could enable this other interface.
